### PR TITLE
add mimalloc (linux only)

### PR DESCRIPTION
**Pull request by elbaro**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [ ] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [ ] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by xhochy** [comment]
What's the reason to use clang here? I thought this should also work with tge default compilers.

**Comment by elbaro** [comment]
Using clang avoided this problem on some machines:
https://github.com/conda/conda-build/issues/3165

Now I applied the better patch, clang is not required anymore.

**Comment by elbaro** [comment]
@conda-forge/staged-recipes Please merge

File: recipes/mimalloc/0001-patch-for-conda.patch [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+diff --git a/src/alloc-override.c b/src/alloc-override.c
+index c0e7bc2..60885a8 100644
+--- a/src/alloc-override.c
++++ b/src/alloc-override.c
+@@ -173,7 +173,7 @@ void* valloc(size_t size)                                   { return mi_valloc(s
+ void* pvalloc(size_t size)                                  { return mi_pvalloc(size); }
+ void* reallocarray(void* p, size_t count, size_t size)      { return mi_reallocarray(p, count, size); }
+ void* memalign(size_t alignment, size_t size)               { return mi_memalign(alignment, size); }
+-void* aligned_alloc(size_t alignment, size_t size)          { return mi_aligned_alloc(alignment, size); }
```
Comment (xhochy) [inline recipes/mimalloc/0001-patch-for-conda.patch:18]: Uh, isn't this breaking the use of `mimalloc` for aligned allocations?

File: recipes/mimalloc/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+source:
+  url: https://github.com/microsoft/mimalloc/archive/v{{ version }}.tar.gz
+  sha256: 7fe9b8f30cd25f8ea35dbb6539652bbf9ea44cb01865bac70a1c368f6b96c770
+  patches:
+    - 0001-patch-for-conda.patch
+
+build:
+  number: 0
+  # If the installation is complex, or different between Unix and Windows, use separate bld.bat and build.sh files instead of this key.
```
Comment (xhochy) [inline recipes/mimalloc/meta.yaml:16]: Please remove the explanative comments from the example.

File: recipes/mimalloc/0001-patch-for-conda.patch (hunk 1 above)
Comment (elbaro) [reply recipes/mimalloc/0001-patch-for-conda.patch:18]: Probably. This is [suggested]((https://github.com/microsoft/mimalloc/issues/219#issuecomment-609880751)) by the mimalloc author, and I am not aware of any other workaround.

**Comment by elbaro** [comment]
@conda-forge/staged-recipes updated.

**Comment by elbaro** [comment]
@xhochy @conda-forge/staged-recipes please review again

**Comment by elbaro** [comment]
How can I ping @conda-forge/staged-recipes (in bold)?

File: recipes/mimalloc/0002-fix-install-path.patch [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+diff --git a/CMakeLists.txt b/CMakeLists.txt
+index e16830a..6afa0b5 100644
+--- a/CMakeLists.txt
++++ b/CMakeLists.txt
+@@ -146,7 +146,7 @@ endif()
+ # Install and output names
+ # -----------------------------------------------------------------------------
+ 
+-set(mi_install_dir "${CMAKE_INSTALL_PREFIX}/lib/mimalloc-${mi_version}")
++set(mi_install_dir "${CMAKE_INSTALL_PREFIX}")
```
Comment (isuruf) [inline recipes/mimalloc/0002-fix-install-path.patch:19]: Can you add a `lib/` at the end?

File: recipes/mimalloc/0002-fix-install-path.patch [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ 
+ # install static and shared library, and the include files
+-install(TARGETS mimalloc EXPORT mimalloc DESTINATION ${mi_install_dir} LIBRARY)
+-install(TARGETS mimalloc-static EXPORT mimalloc DESTINATION ${mi_install_dir})
++install(TARGETS mimalloc EXPORT mimalloc DESTINATION ${mi_install_dir}/lib LIBRARY)
++install(TARGETS mimalloc-static EXPORT mimalloc DESTINATION ${mi_install_dir}/lib)
+ install(FILES include/mimalloc.h DESTINATION ${mi_install_dir}/include)
+ install(FILES include/mimalloc-override.h DESTINATION ${mi_install_dir}/include)
+ install(FILES include/mimalloc-new-delete.h DESTINATION ${mi_install_dir}/include)
+-install(FILES cmake/mimalloc-config.cmake DESTINATION ${mi_install_dir}/cmake)
```
Comment (isuruf) [inline recipes/mimalloc/0002-fix-install-path.patch:34]: Why delete this?

File: recipes/mimalloc/0001-patch-aligned_malloc.patch [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+diff --git a/src/alloc-override.c b/src/alloc-override.c
+index c0e7bc2..cd41147 100644
+--- a/src/alloc-override.c
++++ b/src/alloc-override.c
+@@ -173,7 +173,7 @@ void* valloc(size_t size)                                   { return mi_valloc(s
+ void* pvalloc(size_t size)                                  { return mi_pvalloc(size); }
+ void* reallocarray(void* p, size_t count, size_t size)      { return mi_reallocarray(p, count, size); }
+ void* memalign(size_t alignment, size_t size)               { return mi_memalign(alignment, size); }
+-void* aligned_alloc(size_t alignment, size_t size)          { return mi_aligned_alloc(alignment, size); }
++void* _aligned_malloc(size_t alignment, size_t size) { return mi_aligned_alloc(alignment, size); }
```
Comment (isuruf) [inline recipes/mimalloc/0001-patch-aligned_malloc.patch:19]: Why?

Comment (elbaro) [reply recipes/mimalloc/0001-patch-aligned_malloc.patch:19]: This is the patch from the mimalloc author:
https://github.com/microsoft/mimalloc/issues/219#issuecomment-614220235

File: recipes/mimalloc/0002-fix-install-path.patch (hunk 3 above)
Comment (elbaro) [reply recipes/mimalloc/0002-fix-install-path.patch:19]: That will install under lib/lib.

The current mimalloc realease has a problem with installing path:
https://github.com/microsoft/mimalloc/issues/223

It installs `.o` and `.h` under a wrong path, e.g. `$PREFIX/lib/mimalloc-1.6.1/include/mimalloc_xx.h`

See this test: https://github.com/conda-forge/staged-recipes/pull/11194/files#diff-aee5d9e631c2be348a8e340de9d64b30R37

File: recipes/mimalloc/0002-fix-install-path.patch (hunk 4 above)
Comment (elbaro) [reply recipes/mimalloc/0002-fix-install-path.patch:34]: Reverted.

**Comment by elbaro** [comment]
@isuruf addressed your comments.

**Comment by isuruf** [comment]
Thanks @elbaro for the contribution

**Pull request merged**